vignettes/package.Rmd

```diff
@@ -2,119 +2,250 @@
+Calling `py_require()` is always recommended because it has no effect if the
+user is using a manually managed Python environment, making it generally safe to
+use.
 
-When you do this, you should use the `delay_load` flag to the `import()` function, for example:
+### Typical Usage
 
-``` r
-# global reference to scipy (will be initialized in .onLoad)
-scipy <- NULL
+`py_require()` is typically called from `.onLoad()`, like this:
 
+``` r
 .onLoad <- function(libname, pkgname) {
-  # use superassignment to update global reference to scipy
-  scipy <<- reticulate::import("scipy", delay_load = TRUE)
+  reticulate::py_require("scipy")
 }
 ```
 
-Using the `delay_load` flag has two important benefits:
+It can also be called from other package functions. For example, the `keras3` R
+package defines a `use_backend()` function that calls `py_require()` with
+different values based on the user-requested backend:
 
-1.  It allows you to successfully load your package even when Python / Python packages are not installed on the target system (this is particularly important when testing on CRAN build machines).
+``` r
+.onLoad <- function(...) {
+  py_require(c("keras", "tensorflow")) # default tensorflow backend
+}
 
-2.  It allows users to specify a desired location for Python before interacting with your package. For example:
+use_backend <- function(backend, gpu = TRUE) {
+  py_require("tensorflow", action = "remove") # remove default backend
+
+  switch(paste0(backend, "_", get_os()),
+    Linux_jax = if (gpu) py_require("jax[cuda12]") else py_require("jax[cpu]"),
+    macOS_jax = py_require(c("jax", if (gpu) "jax-metal")),
+    Windows_jax = py_require("jax"),
+    Linux_tensorflow = { ... },
+    macOS_tensorflow = { ... },
+    Windows_tensorflow = { ... },
+    Linux_torch = { ... },
+    macOS_torch = { ... },
+    Windows_torch = { ... }
+  )
+}
+```
 
-    ``` r
-    library(mypackage)
-    reticulate::use_virtualenv("~/pythonenvs/userenv")
-    # call functions from mypackage
-    ```
+keras3 users can then specify a backend like this:
+
+``` r
+library(keras3)
+use_backend("jax")
+```
+
+### Best Practices
+
+Be mindful that other R packages and users may also declare Python requirements.
+Avoid restrictive version constraints. If a version constraint is necessary,
+prefer `>=` and `!=` over `<=`, as the latter can quickly become outdated.
+
+An example user script header:
+
+``` r
+library(pysparklyr) # declares requirements for PySpark
+library(keras3)     # declares requirements for default 'tensorflow' backend
+use_backend("jax")  # removes 'tensorflow' requirements, adds 'jax' requirements
+
+library(reticulate)
+py_require(c("scipy", "polars"))         # user-declared requirements
+py_require(python_version = ">=3.12")
+py_require(exclude_newer = "2025-01-19")
```

Comment:
What does this do? `exclude_newer` hasn't been discussed in this vignette; is it needed?

Reply:
`exclude_newer` is discussed in the other vignette. I included it here as a reminder to package authors that their requirements will be combined with a wider variety of user requirements. If you think it's too confusing here I can remove it.